src/input/agent/cost_limit.rs

```diff
@@ -0,0 +1,199 @@
+        AgentCostLimits {
+            capex_limit: self.capex_limit,
+            annual_cost_limit: self.annual_cost_limit,
+        }
+    }
+}
+
+/// Read agent cost limits info from the agent_cost_limits.csv file.
+///
+/// # Arguments
+///
+/// * `model_dir` - Folder containing model configuration files
+///
+/// # Returns
+///
+/// A map of Agents, with the agent ID as the key
+pub fn read_agent_cost_limits(
+    model_dir: &Path,
+    agent_ids: &HashSet<AgentID>,
+    milestone_years: &[u32],
+) -> Result<HashMap<AgentID, AgentCostLimitsMap>> {
+    let file_path = model_dir.join(AGENT_COST_LIMITS_FILE_NAME);
+    let agent_cost_limits_csv = read_csv(&file_path)?;
+    read_agent_cost_limits_from_iter(agent_cost_limits_csv, agent_ids, milestone_years)
+        .with_context(|| input_err_msg(&file_path))
+}
+
+fn read_agent_cost_limits_from_iter<I>(
+    iter: I,
+    agent_ids: &HashSet<AgentID>,
+    milestone_years: &[u32],
+) -> Result<HashMap<AgentID, AgentCostLimitsMap>>
+where
+    I: Iterator<Item = AgentCostLimitsRaw>,
+{
+    let mut map: HashMap<AgentID, AgentCostLimitsMap> = HashMap::new();
+    for agent_cost_limits_raw in iter {
+        let cost_limits = agent_cost_limits_raw.to_agent_cost_limits();
+        let years = parse_year_str(&agent_cost_limits_raw.year, milestone_years)?;
+
+        // Get agent ID
+        let agent_id = agent_ids.get_id_by_str(&agent_cost_limits_raw.agent_id)?;
+
+        // Get or create entry in the map
+        let entry = map.entry(agent_id.clone()).or_default();
+
+        // Insert cost limits for the specified year(s)
+        for year in years {
+            entry.insert(year, cost_limits.clone());
+        }
+    }
+
+    // Validation: if cost limits are specified for an agent, they must be present for all years.
+    for (id, cost_limits) in map.iter() {
+        for year in milestone_years {
+            if !cost_limits.contains_key(year) {
+                return Err(anyhow::anyhow!(
+                    "Agent {} is missing cost limits for year {}",
+                    id,
+                    year
+                ));
+            }
+        }
+    }
+
+    Ok(map)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::agent::AgentCostLimits;
+    use std::collections::HashSet;
+
+    fn create_agent_cost_limits_raw(
+        agent_id: &str,
+        year: &str,
+        capex_limit: Option<f64>,
+        annual_cost_limit: Option<f64>,
+    ) -> AgentCostLimitsRaw {
+        AgentCostLimitsRaw {
+            agent_id: agent_id.to_string(),
+            year: year.to_string(),
+            capex_limit,
+            annual_cost_limit,
+        }
+    }
+
+    #[test]
+    fn test_read_agent_cost_limits_from_iter_all_years() {
+        let agent_ids: HashSet<AgentID> = ["Agent1", "Agent2"]
+            .iter()
+            .map(|&id| AgentID::from(id))
+            .collect();
+        let milestone_years = vec![2020, 2025];
+
+        let iter = vec![
```

Comment:
I think these can just be slices:

```suggestion
        let milestone_years = [2020, 2025];

        let iter = [
```